ConfigSpace/forbidden.pyx

```diff
@@ -564,13 +568,74 @@ cdef class ForbiddenRelation(AbstractForbiddenComponent):
             else:
                 return False
 
-        # Relation is always evaluated against actual value and not vector representation
-        return self._is_forbidden(self.left._transform(left), self.right._transform(right))
+        # Relation is always evaluated against actual value and
+        # not vector representation
+        return self._is_forbidden(self.left._transform(left),
+                                  self.right._transform(right))
 
     cdef int _is_forbidden_vector(self, DTYPE_t left, DTYPE_t right) except -1:
         pass
 
 
+cdef class ForbiddenCallableRelation(ForbiddenRelation):
+    """A ForbiddenCallable relation between two hyperparameters.
+
+    The ForbiddenCallable uses two hyperparameters as input to a
+    specified callable, which returns True if the relationship
+    between the two hyperparameters is forbidden.
+
+    >>> from ConfigSpace import ConfigurationSpace, ForbiddenLessThanRelation
+    >>>
+    >>> cs = ConfigurationSpace({"a": [1, 2, 3], "b": [2, 5, 6]})
+    >>>
+    >>> forbidden_clause = ForbiddenCallableRelation(cs['a'], cs['b'])
+    >>> cs.add_forbidden_clause(forbidden_clause)
+    Forbidden: f(a,b) == True
+
+    Parameters
+    ----------
+     left : :ref:`Hyperparameters`
+         first argument of callable
+
+     right : :ref:`Hyperparameters`
+         second argument of callable
+
+     f : A callable that relates the two hyperparameters
+    """
+    cdef public f
+
+    def __init__(self, left: Hyperparameter, right: Hyperparameter,
+                 f: ForbiddenCallable):
+        if not isinstance(f, Callable):  # Can't use ForbiddenCallable here apparently
+            raise TypeError("Argument 'f' is not of type %s." % Callable)
+
+        super().__init__(left, right)
+        self.f = f
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, self.__class__):
+            return False
+        return super().__eq__(other) and self.f == other.f
+
+    def __copy__(self):
+        return self.__class__(
+            a=copy.copy(self.left),
+            b=copy.copy(self.right),
+            f=copy.copy(self.f)
+        )
+
+    def __repr__(self):
+        from inspect import getsource
+        f_source = getsource(self.f)
+        return f"Forbidden: {f_source} | Arguments: {self.left.name}, {self.right.name}"
```

Comment:
This `getsource` call could be quite large and rather verbose. For example:

```python
def f():
    print("hello")
    x = 1 + 2
    y = "hi" + "mars"
    return b"no"
    
from inspect import getsource
getsource(f)
# 'def f():\n    print("hello")\n    x = 1 + 2\n    y = "hi" + "mars"\n    return b"no"\n'
```

I think we could just get around this by using the functions, `f.__qualname__` instead.

Reply:
As a follow up you could accomodate lambdas as:

```python
qualname = f.__qualname__
f_repr = getsource(f) if qualname == "<lambda>" else qualname
```

Reply:
That's a good idea. I was noticing formatting issues using `getsource` as well since it keeps all of the leading white space.

Reply:
Obtaining only the source code of the lambda (and discarding any other code on the same line) is apparently non-trivial. See for example https://stackoverflow.com/questions/59498679/how-can-i-get-exactly-the-code-of-a-lambda-function-in-python. I have revised it to use the qualname though.